### ISSUE TITLE
Question Regarding Loss Computation in RAGEN

### ISSUE THREAD

**Issue by Mr-Potential** [RAGEN-AI/RAGEN#54]
Greetings, I have some inquiries regarding the issue of loss calculation that I wish to consult with you about.

RAGEN calculate log_prob for both responses and observation(obs is continuously concatenated to right side response in generation.py). However obs is not generated from our LLM, will gradient calculated from obs tokens lead to some unexpected result? Suppose our obs is very large, will this algorithm fail due to this issue?

```
def _update_right_side(self, right_side: Dict,
                       cur_responses: torch.Tensor,
                       next_obs_ids: torch.Tensor) -> Dict:
    """Update right side state."""
    responses = self.tensor_fn.concatenate_with_padding([
        right_side['responses'],
        cur_responses,
        next_obs_ids
    ], pad_to_left=False)

    effective_len = self.tensor_fn.create_attention_mask(responses).sum(dim=1).max()
    max_len = min(self.config.max_prompt_length, effective_len)

    return {'responses': responses[:, :max_len]}
```

**Comment by Mr-Potential** [comment]
It seems your partner JamesKrW has already fixed this issue in previous commits last month(my bad I didn't notice it when I first looked into your project), however you asked me to check jedward225's progress, I am a little bit confused here, it seems their works are duplicated, may I know what's the difference?

![Image](https://github.com/user-attachments/assets/aec2caac-0463-445a-8279-258f2b1b4725)

**Comment by jedward225** [comment]
You're right, in my point of view, calculating gradients for observation tokens is problematic because:

Observations come from the environment, not the LLM, so large observation sequences can dominate the loss calculation, leading to unstable training

I'm actually trying to implement an "observation masking" solution. The approach in my draft is:

1. Distinguishes between model-generated content (actions/thinking) and environment-provided content (observations)
2. Creates a loss mask that includes only tokens the model generates
3. Applies this mask during training to exclude observation tokens from gradient computation

My implementation in `_create_loss_mask` creates separate masks for observations vs. actions/thinking, and only includes the latter in loss calculation. This prevents the model from receiving gradients for tokens it didn't generate.

As for JamesKrW's work, I'm not quite sure about the answer. But, emmmm, his update may not be the solution to ur question, I think.

**Comment by Mr-Potential** [comment]
Thanks for your swift response! I will be looking into your draft soon. May I ask if you finalize your update after running some tests? I found exploding grad norm issue when I try settings with larger obs sequences, did you notice whether your updates can mitigate this issue?

**Comment by jedward225** [comment]
Actually that's what I'm doing these 2 days, I had some trouble with my server so got some delay. But I'll immediately run some tests and tell you if it works or not. Thanks again for your questions and attention to this project : >

**Comment by jedward225** [comment]
problem solved, a pr would be merged these days(actually the bug is located in `_create_loss_mask`. previous implementation mistook the meaning of `0` token and `1` token)

**Comment by ZihanWang314** [comment]
Hi, just to let you know that we've updated a new version of code with improved stability training stability, and we wrote a [paper](https://github.com/RAGEN-AI/RAGEN/blob/main/RAGEN.pdf). The loss mask can now be turned on within the `config/base.yaml` file, with the key as `enable_response_mask`. Thanks!
